sdk/storage/azure-storage-blob/src/main/java/com/azure/storage/blob/BlobAsyncClient.java

```diff
@@ -287,76 +289,145 @@ private SpecializedBlobClientBuilder prepareBuilder() {
                 ? new ParallelTransferOptions(null, null, null)
                 : new ParallelTransferOptions(parallelTransferOptions.getBlockSize(),
-                    parallelTransferOptions.getNumBuffers(), parallelTransferOptions.getProgressReceiver());
-
-            // See ProgressReporter for an explanation on why this lock is necessary and why we use AtomicLong.
-            AtomicLong totalProgress = new AtomicLong(0);
-            Lock progressLock = new ReentrantLock();
-
-            // Validation done in the constructor.
-            UploadBufferPool pool = new UploadBufferPool(finalParallelTransferOptions.getNumBuffers(),
-                finalParallelTransferOptions.getBlockSize());
-
-            /*
-            Break the source Flux into chunks that are <= chunk size. This makes filling the pooled buffers much easier
-            as we can guarantee we only need at most two buffers for any call to write (two in the case of one pool
-            buffer filling up with more data to write). We use flatMapSequential because we need to guarantee we
-            preserve the ordering of the buffers, but we don't really care if one is split before another.
-             */
-            Flux<ByteBuffer> chunkedSource = data
-                .filter(ByteBuffer::hasRemaining)
-                .flatMapSequential(buffer -> {
-                    if (buffer.remaining() <= finalParallelTransferOptions.getBlockSize()) {
-                        return Flux.just(buffer);
-                    }
-                    int numSplits =
-                        (int) Math.ceil(buffer.remaining() / (double) finalParallelTransferOptions.getBlockSize());
-                    return Flux.range(0, numSplits)
-                        .map(i -> {
-                            ByteBuffer duplicate = buffer.duplicate().asReadOnlyBuffer();
-                            duplicate.position(i * finalParallelTransferOptions.getBlockSize());
-                            duplicate.limit(Math.min(duplicate.limit(),
-                                (i + 1) * finalParallelTransferOptions.getBlockSize()));
-                            return duplicate;
-                        });
-                });
+                parallelTransferOptions.getNumBuffers(), parallelTransferOptions.getProgressReceiver());
+
+            BlockBlobAsyncClient blockBlobAsyncClient = getBlockBlobAsyncClient();
+            return determineUploadFullOrChunked(data, (stream) -> uploadInChunks(blockBlobAsyncClient, stream,
+                finalParallelTransferOptions, headers, metadata, tier, accessConditionsFinal),
+                (stream, length) -> blockBlobAsyncClient.uploadWithResponse(
+                    ProgressReporter.addProgressReporting(stream, finalParallelTransferOptions.getProgressReceiver()),
+                    length, headers, metadata, tier, null, accessConditionsFinal));
```

Comment:
Nit. Please use better indentation, as there's a lot of nested calls in here.

Reply:
Done